### PR TITLE
Customizable PDA ringtones

### PR DESCRIPTION
**Pull request by Superlagg**
[FEATURE]
## About the PR 

PDAs can now have their ringtones changed. They are stored in datums and applied through various programs. 

The ringtone datums are highly customizable. You're able to define any number of sounds, whether they're played from that list in order or randomly, the playsound vars for each sound, specific ringer messages to use it set to use them, actions for the ringtone to take after playing certain ringtones, and the ability for the ringtone to listen to your text messages for instructions.

The ringtones can be previewed on your PDA by pressing the preview button. It'll set itself as your PDA's temporary ringtone and send you a message.

Adds the following ringtones:

- The stock standard Two-Beep, available on all PDAs and on a cart in the vendor.
- An obnoxious "wolf pack" cart with random dog noises. Available for about 800,000 credits.
- Some dopey elementary school not-arg learning cart lodged in a monkey sometimes.
- A bundle of silly noises that come with the clown's cart.
- A pirated sound effects library for traitors that'll play explosion noises, gunfire, and csaber stuff while messing up your PDA's speakers.

Also lets you block specific mailgroups, alerts, and people.

Also adds automatic filesharing, where you can host a file on your PDA, and it'll send that file to anyone who messages you a specific passcode. You need the SpaceWire filesharing program to do this, and it can work with mailgroups to send a message to those groups with the passcode, so you can send a file to everyone in your department. Currently only spawns on head and QM PDAs.

## Why's this needed? 

PDAs need more options to be obnoxious. Also I like the idea of each department getting their own ringtones

## Changelog

```
(u)Superlagg:
(*)PDAs can now use different ringtones! Ask the clown if you can borrow their ringtone program.
(*)PDAs can now block specific people, mailgroups, and alerts.
(+)The Syndicate are dirty software pirates, supplying their agents with stolen copies of a certain PDA compatible sound effect library for a single TC.
(+)The heads of staff have a new filesharing program, SpaceWire, which lets them automatically share PDA files to anyone who messages them a secret password.
```

### CODE REVIEW

**Comment by Superlagg** [comment]
You can now set ringtones for specific mail groups and alert types.

**Comment by Superlagg** [comment]
Added block functionality to mailgroups, alerts, and specific people.

Also added automatic filesharing, where you can host a file on your PDA, and it'll send that file to anyone who messages you a specific passcode. You need the SpaceWire filesharing program to do this, and it can work with mailgroups to send a message to those groups with the passcode, so you can send a file to everyone in your department. Currently only spawns on head and QM PDAs.

File: code/obj/item/device/pda2/base_os.dm [hunk 1]
```diff
@@ -733,7 +800,20 @@
 							if (src.master)
 								src.master.explode()
 
-					src.master.display_alert(alert_beep)
+					if(senderName in src.blocked_numbers)
+						return
+
+					var/previewtext = (signal.data["tag"] == "preview_message")
+					if(previewtext && ON_COOLDOWN(src, "preview_cooldown", 15 SECONDS))
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:807]: The lack of feedback about preview cooldown made me think the feature is broken at first. At least throw a boutput here with a message about the cooldown or something like that.

File: code/obj/item/device/pda2/base_os.dm [hunk 2]
```diff
@@ -755,6 +835,12 @@
 					var/senderassignment = signal.data["sender_assignment"]
 					var/file_ext = signal.data["file_ext"]
 					var/filesize = signal.data["file_size"]
+					var/signalTag = signal.data["tag"]
+					var/signalGroup = signal.data["group"]
+
+					if(signalGroup) //Check to see if we have this ~mailgroup~
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:841]: Doesn't this clash with #3178 which allows for groups to be a list?

File: code/obj/item/device/pda2/base_os.dm [hunk 3]
```diff
@@ -696,24 +758,29 @@
 						else if (groupAddress in src.master.muted_mailgroups)
 							return
 
-					var/sender = signal.data["sender_name"]
-					if(!sender)
-						sender = "!Unknown!"
+					var/signalTag = signal.data["tag"]
+					if(signalTag in src.muted_alerts)
+						return
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:763]: Does this PR move the system which marks messages as alerts to tags? Just making sure.

Comment (pali6) [reply code/obj/item/device/pda2/base_os.dm:763]: What about possibility of multiple tags?

File: code/obj/item/device/pda2/base_os.dm [hunk 4]
```diff
@@ -292,6 +287,75 @@
...
+
+						for(var/alert in src.master.alertgroups)
+							var/datum/ringtone/rt = null
+							var/rtButton = "Default"
+							var/muteButton = "<a href='byond://?src=\ref[src];manageBlock=["add"];type=["alert"];entry=[alert]'>Mute</a>"
+							if(istype(src.master.alert_ringtones[alert], /datum/ringtone))
+								rt = src.master.alert_ringtones[alert]
+								rtButton = "[rt.name]"
+							if(alert in src.muted_alerts)
+								muteButton = "<a href='byond://?src=\ref[src];manageBlock=["remove"];type=["alert"];entry=[alert]'>Unblock</a>"
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:328]:
```suggestion
								muteButton = "<a href='byond://?src=\ref[src];manageBlock=["remove"];type=["alert"];entry=[alert]'>Unmute</a>"
```

File: code/obj/item/device/pda2/ringtone.dm [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+	/// Include an option to override the PDA's alert message?
+	var/overrideAlertAllowed = 1
+	/// Should this ringtone's own alert message overwrite the PDA's?
+	var/overrideAlertMessage = 0
+	/// The text that asks the user whether to override the alert messages or not
+	var/overrideAlertText = "Use default ringer message?"
+	/// The text that indicates an affirmative for overriding the alert text
+	var/overrideAlertYesText = "Yes"
+	/// The text that indicates a negative for overriding the alert text
+	var/overrideAlertNoText = "No"
```
Comment (pali6) [inline code/obj/item/device/pda2/ringtone.dm:24]: Most of the funny overrides of the buttons are fine but imo this one's not really intuitive in what it does even when explained clearly and obscuring it with the overrides makes it reall confusing. I think I'd want this to not be overriden.

**Review by pali6 — CHANGES_REQUESTED** [review]
Very incomplete review, I only looked through like 15% of the code but also it's bedtime. Expect more reviews later on.

Muting alerts doesn't seem to work anymore from my testing.

I have some concerns about possible annoyance impact of this PR as a whole but I think overall it's good and if it proves too annoying we can do something about it later.

File: code/obj/item/device/pda2/base_os.dm (hunk 1 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/base_os.dm]: Done! Moved the check to the ringtone program and made it output a message via the PDA about why the button isn't working.

File: code/obj/item/device/pda2/base_os.dm (hunk 3 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/base_os.dm]: Originally, I had mailgroups identified via "group", and alerts via "tag", as previously there could be only one of each. But now that they can have multiple groups assigned to the same signal, it made more sense just to lump everything into the mailgroup check. And I forgot to change it until now. Should support multiple tags now.

File: code/obj/item/device/pda2/base_os.dm (hunk 2 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/base_os.dm]: Not anymore it doesnt!

File: code/obj/item/device/pda2/ringtone.dm (hunk 5 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/ringtone.dm]: I rewrote some of the prompts to be more descriptive with what it does. I still need it to be overridable, as not every alert message override is itself a typical PDA alert message.

**Comment by Superlagg** [comment]
> Very incomplete review, I only looked through like 15% of the code but also it's bedtime. Expect more reviews later on.
> 
> Muting alerts doesn't seem to work anymore from my testing.

I forgot to update alert muting to use the mailgroup mute list. Should work now.

> I have some concerns about possible annoyance impact of this PR as a whole but I think overall it's good and if it proves too annoying we can do something about it later.

Part of the annoyance is intended, mainly from the fact that your coworker has an obnoxious ringtone. Hopefully it won't be *too* annoying! If needed, I could add a way to break a stolen PDA's speakers.

**Comment by pali6** [comment]
Changing alert ringtone doesn't change alert ringtone.

**Comment by pali6** [comment]
Not strictly necessary but it'd be neat if you could see currently set ringtones somewhere and maybe use that menu to revert back to default? Or maybe even use that menu to set one of the ringtones from the list as an alternative way of setting ringtones.

**Comment by Superlagg** [comment]
> Not strictly necessary but it'd be neat if you could see currently set ringtones somewhere and maybe use that menu to revert back to default? Or maybe even use that menu to set one of the ringtones from the list as an alternative way of setting ringtones.

Currently set ringtones should show up in the groups menu and reset the ringtone if clicked, unless I messed something up.

Using that menu to select ringtones would be tricky, though. I could try having the ringtone programs "install" themselves to a list of stored ringtones, then set them from there... but that'd likely need to be for a later PR.

File: code/datums/computerfiles.dm [hunk 6]
```diff
@@ -107,6 +107,8 @@
 
 
 	file
+		/// Prevent it from piracy via PDA filehosting?
+		var/dont_copy = 0
```
Comment (pali6) [inline code/datums/computerfiles.dm:111]: Imo trying to copy a program with `dont_copy == 1` should again produce an error message. Or the `Copy` button could just be grayed out for these.

File: code/obj/item/device/pda2/base_os.dm [hunk 7]
```diff
@@ -292,6 +285,75 @@
...
+					. += "<a href='byond://?src=\ref[src];mode=[MODE_MESSAGE]'>Back</a><br>"
+					. += "<h4>Address Book</h4><br>"
+					if(length(src.all_callers) < 1)
+						. += "Address book is empty!"
+					else
+						. += "<table cellspacing=5>"
+						for(var/caller in src.all_callers)
+							var/muteButton = "<a href='byond://?src=\ref[src];manageBlock=["add"];type=["single"];entry=[src.all_callers[caller]]'>Block</a>"
+							var/callButton = "<a href='byond://?src=\ref[src];input=message;target=[caller]'>Msg</a>"
+							var/sendButton = "<a href='byond://?src=\ref[src];input=send_file;target=[caller]'>Send</a>"
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:340]:
```suggestion
							var/sendButton = "<a href='byond://?src=\ref[src];input=send_file;target=[caller]'>Send File</a>"
```

File: code/obj/item/device/pda2/base_os.dm [hunk 8]
```diff
@@ -292,6 +285,75 @@
...
+							var/delButton = "<a href='byond://?src=\ref[src];delAddress=[caller]'>Del</a>"
+							if(caller in src.blocked_numbers)
+								muteButton = "<a href='byond://?src=\ref[src];manageBlock=["remove"];type=["single"];entry=[src.all_callers[caller]]'>Unblock</a>"
```
Comment (pali6) [inline code/obj/item/device/pda2/base_os.dm:343]: Not sure why but this doesn't work.

File: code/obj/item/device/pda2/base_os.dm (hunk 7 above)
Comment (pali6) [reply code/obj/item/device/pda2/base_os.dm:340]: Also this should imo be grayed out if there's no file in the clipboard or if the file in the clipboard is not copyable.

**Review by pali6 — CHANGES_REQUESTED** [review]
Another partial review: As mentioned above alert ringtones don't work and it's not possible to reset group/alert ringtones to default.

File: code/datums/computerfiles.dm (hunk 6 above)
Comment (Superlagg) [reply code/datums/computerfiles.dm:111]: Good idea! If a file can't be copied, the link is replaced with a strikethrough, now.

File: code/obj/item/device/pda2/base_os.dm (hunk 7 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/base_os.dm]: Done!

File: code/obj/item/device/pda2/base_os.dm (hunk 8 above)
Comment (Superlagg) [reply code/obj/item/device/pda2/base_os.dm]: It was comparing an address to a list of names. Now, it compares the right values and should work just fine!

File: code/datums/pda_ringtones.dm [hunk 9]
```diff
@@ -0,0 +1,748 @@
...
+	applyText = "Apply"
+	previewText = "Sample"
+	nameText = "SounPacK:"
+	descText = "DescrIptioN:"
+
+	DoSpecialThing(var/index)
+		animate_shockwave(src.holder)
+		if(index == 6) return
+		for(var/client/C in clients)
+			if(C.mob && (C.mob.z == src.holder?.loc?.z))
```
Comment (pali6) [inline code/datums/pda_ringtones.dm:628]:
```suggestion
		var/turf/T = get_turf(src.holder)
		for(var/client/C in clients)
			if(C.mob && (C.mob.z == T?.z))
```

**Review by pali6 — CHANGES_REQUESTED** [review]

**Comment by pali6** [comment]
I haven't tested the filesharing feature but someone will probably scream if it's broken.

**Review by pali6 — APPROVED** [review]

**Pull request merged**